crypto/cipher_extra/e_chacha20poly1305.c

```diff
@@ -346,3 +382,317 @@ const EVP_AEAD *EVP_aead_chacha20_poly1305(void) {
 const EVP_AEAD *EVP_aead_xchacha20_poly1305(void) {
   return &aead_xchacha20_poly1305;
 }
+
+static int cipher_chacha20_poly1305_init_key(CIPHER_CHACHA_POLY_CTX *ctx,
+                                 const uint8_t user_key[CHACHA_KEY_LEN],
+                                 const uint8_t counter_nonce[CHACHA_CTR_IV_LEN])
+{
+  CIPHER_CHACHA_KEY *key = CC_KEY(ctx);
+  uint32_t i;
+  if (user_key) {
+    for (i = 0; i < (CHACHA_KEY_LEN / 4); i++) {
+      key->key[i] = *(((const uint32_t *) user_key) + i);
+    }
+  }
+  if (counter_nonce) {
+    for (i = 0; i < CHACHA_CTR_IV_LEN / 4; i++) {
+      key->counter_nonce[i] = *(((const uint32_t *) counter_nonce) + i);
+    }
+  }
+  key->partial_len = 0;
+  return 1;
+}
+
+static int cipher_chacha20_poly1305_init(EVP_CIPHER_CTX *ctx,
+                                         const uint8_t *key,
+                                         const uint8_t *iv, int32_t enc) {
+  CIPHER_CHACHA_POLY_CTX *cipher_ctx = CCP_CTX(ctx);
+  if (!key && !iv) {
+    return 1;
+  }
+  cipher_ctx->len.aad = 0;
+  cipher_ctx->len.text = 0;
+  cipher_ctx->pad_aad = 0;
+  cipher_ctx->poly_initialized = 0;
+  // Init can be called multiple times before starting the cipher to
+  // independently initialize any combination of Key/IV/NULL.
+  if (iv != NULL) {
+    // Start the counter at 0 and copy over the nonce(iv)
+    uint8_t counter_nonce[CHACHA_CTR_IV_LEN] = {0};
+    memcpy(counter_nonce + CHACHA_CTR_IV_LEN - CHACHA_IV_LEN, iv,
+           CHACHA_IV_LEN);
+    cipher_chacha20_poly1305_init_key(cipher_ctx, key, counter_nonce);
+    // Nonce occupies the last 3 indices of the array
+    cipher_ctx->iv[0] = cipher_ctx->key.counter_nonce[1];
+    cipher_ctx->iv[1] = cipher_ctx->key.counter_nonce[2];
+    cipher_ctx->iv[2] = cipher_ctx->key.counter_nonce[3];
+  } else {
+    cipher_chacha20_poly1305_init_key(cipher_ctx, key, NULL);
+  }
+  return 1;
+}
+
+static int cipher_chacha20_do_cipher(EVP_CIPHER_CTX *ctx, uint8_t *out,
+                                     const uint8_t *inp, size_t len)
+{
+  CIPHER_CHACHA_POLY_CTX  *cipher_ctx = CCP_CTX(ctx);
+  CIPHER_CHACHA_KEY *key = CC_KEY(cipher_ctx);
+  uint32_t n, rem, ctr32;
+
+  // Complete any partial block
+  if ((n = key->partial_len)) {
```

Comment:
Personally I prefer separating assignment out of any conditional blocks to make it easier to read/reason about.